test/unit/datasets/test_datasets.py

```diff
@@ -38,3 +40,26 @@ def test_dataset_as_X_y(self, as_frame, fetch_function):
         assert isinstance(X, pd.DataFrame if as_frame else np.ndarray)
         assert y is not None
         assert isinstance(y, pd.Series if as_frame else np.ndarray)
+
+    def test_synthetic_datasets(self):
```

Comment:
Can we add the following test cases as well?
- pass in custom configuration with a different number of samples than the default and assert that the sum of the `n_samples` values adds up to the total size
- pass in a different number of features and assert the resulting array dimension

Bonus: try out invalid args
- `n_features=0`
- empty dict for `classes`

Reply:
As it stands, such invalid args result in a `ValueError`. Is that acceptable, or should we do some kind of sanitization for the user? I'm specifically thinking about `classes` (now called `feature_config`). It's not very user-friendly to ask for this dict in the first place, and I hope we can replace it with something cleaner in the follow-up, but for now I wonder if we should try to ensure that the dict has entries of the form `'label': {'n_samples': n_samples, 'class_sep': class_sep}`.

Reply:
That's fine 🙂 We have plenty of tests for invalid inputs where we assert the correct errors/exceptions are raised. Typically comes out as
```
with pytest.raises(ValueError) as error:
    assert str(error.value) == expected_msg
```

I completely agree on the dict not being ideal. I started thinking a bit about having a well-defined config object to which you can add groups. Perhaps something that would work like this:
```
sensitive_feature_config = SensitiveFeatureConfig(names=["gender", "age"])
sensitive_feature_config.add_group(gender="male", "age"="0-20", n_samples=123, class_sep=1.1, ...)
sensitive_feature_config.add_group(gender="female", "age"="0-20", n_samples=321, class_sep=0.6, ...)
sensitive_feature_config.add_group(gender="non-binary", "age"="0-20", n_samples=231, class_sep=1.7, ...)
...
sensitive_feature_config.add_group(gender="male", "age"="80+", n_samples=103, class_sep=1.05, ...)
sensitive_feature_config.add_group(gender="female", "age"="80+", n_samples=183, class_sep=1.4, ...)
sensitive_feature_config.add_group(gender="non-binary", "age"="80+", n_samples=123, class_sep=0.2, ...)
...
```
[I would allow all sorts of args to be passed through to `make_classification` via `kwargs`.]

Again, I think this goes quite a bit beyond the basic version you have in this PR, so I wouldn't expect this right away, and I feel like we should get some more input on this from others such as @adrinjalali @MiroDudik @riedgar-ms @hildeweerts

Reply:
This seems like a nice way to handle multiple features. I hadn't thought of dealing directly with group intersections in the config.

Reply:
I like @romanlutz 's proposal, except that I'd have an object for each sensitive feature, and not have the Cartesian product written in the code.

I think it's worth having the API as something we like in the first version. I rather not have to deprecate later since we already know we don't like the API as is. I'm happy to have a `SensitiveFeatureGenerator` object, but the API needs to be a bit less verbose than what's proposed by Roman. I'm not sure if we have to give users this much control over how the data is generated. If they really need this much control, they can write their own generator. This function is supposed to generate some dummy data for demonstration purposes. It's not supposed to generate data similar to the real world issues we have.

Reply:
I've tried something like this here: coreysharris/fairlearn#1. I opened it as a PR against this branch so as not to muddy the waters here if people don't like it. I think the API is much cleaner, and it makes it easier to extend functionality later.

One note: although it _looks_ like it now supports multiple features, that's not the case. The implementation needs to be reworked, but I haven't gotten that far.

Reply:
Somehow I missed this, sorry for the delay!
I **love** it! Do you want to pull that change into your branch so that we can review it as part of this PR?

Reply:
Ok, great! In order to support multiple features, the API has to change slightly. It looks more and more like your suggestion above 😛